src/blueprint.zig

```diff
@@ -372,6 +374,228 @@ pub const Pattern = struct {
 	}
 };
 
+fn generatePropertyEnum() type {
+	var tempFields: [@typeInfo(Block).@"struct".decls.len]std.builtin.Type.EnumField = undefined;
+	var count = 0;
+
+	for(std.meta.declarations(Block)) |decl| {
+		const declInfo = @typeInfo(@TypeOf(@field(Block, decl.name)));
+		if(declInfo != .@"fn") continue;
+		if(declInfo.@"fn".return_type != bool) continue;
+		if(declInfo.@"fn".params.len != 1) continue;
+
+		tempFields[count] = .{.name = decl.name, .value = count};
+		count += 1;
+	}
+
+	const outFields: [count]std.builtin.Type.EnumField = tempFields[0..count].*;
+
+	return @Type(.{.@"enum" = .{
+		.tag_type = u8,
+		.fields = &outFields,
+		.decls = &.{},
+		.is_exhaustive = true,
+	}});
+}
+
+pub const Mask = struct {
+	const AndList = ListUnmanaged(Entry);
+	const OrList = ListUnmanaged(AndList);
+
+	entries: OrList,
+
+	const or_ = '|';
+	const and_ = '&';
+	const inverse = '!';
+	const tag = '$';
+	const property = '@';
+
+	const Entry = struct {
+		inner: Inner,
+		isInverse: bool,
+
+		const Inner = union(enum) {
+			block: Block,
+			blockType: u16,
+			blockTag: Tag,
+			blockProperty: Property,
+
+			const Property = generatePropertyEnum();
+
+			fn initFromString(specifier: []const u8) !Inner {
+				return switch(specifier[0]) {
+					tag => .{.blockTag = Tag.get(specifier[1..]) orelse return error.TagNotFound},
+					property => .{.blockProperty = std.meta.stringToEnum(Property, specifier[1..]) orelse return error.PropertyNotFound},
+					else => return try parseBlockLike(specifier),
+				};
+			}
+
+			fn match(self: Inner, block: Block) bool {
+				return switch(self) {
+					.block => |desired| block.typ == desired.typ and block.data == desired.data,
+					.blockType => |desired| block.typ == desired,
+					.blockTag => |desired| block.hasTag(desired),
+					.blockProperty => |blockProperty| switch(blockProperty) {
+						inline else => |prop| @field(Block, @tagName(prop))(block),
+					},
+				};
+			}
+		};
+
+		fn initFromString(specifier: []const u8) !Entry {
+			const isInverse = specifier[0] == '!';
+			const entry = try Inner.initFromString(specifier[if(isInverse) 1 else 0..]);
+			return .{.inner = entry, .isInverse = isInverse};
+		}
+
+		pub fn match(self: Entry, block: Block) bool {
+			const isMatch = self.inner.match(block);
+			if(self.isInverse) {
+				return !isMatch;
+			}
+			return isMatch;
+		}
+	};
+
+	pub fn initFromString(allocator: NeverFailingAllocator, source: []const u8) !@This() {
+		var orStorage: OrList = .{};
+		errdefer orStorage.deinit(allocator);
```

Comment:
This is leaking memory if the error happens not on the first iteration. Instead you'd need to have a loop here that also frees all the `and` lists. At this point, I'd suggest to just use
```
var result: @This() = .{.entries = .{}};
errdefer result.deinit();
```